tests/semantic_analyser.cpp

```diff
@@ -4035,4 +4035,78 @@ kfunc:foo { $p = path((uint8 *)0) }
   test("BEGIN { $b = buf(0, 10) }");
 }
 
+TEST(semantic_analyser, variable_declarations)
+{
+  test("BEGIN { let $a; $a = 1; }");
+  test("BEGIN { let $a: int16; $a = 1; }");
+  test("BEGIN { let $a = 1; }");
+  test("BEGIN { let $a: int16 = 1; }");
+  test("BEGIN { let $a: str_t[5] = \"hiya\"; }");
+  // If the type is specified it's strict in that future assignments
+  // need to fit into that type
+  test(R"(BEGIN { let $a: str_t[5] = "hiya"; $a = "bye"; })");
+  test(R"(BEGIN { let $a = "hiya"; $a = "longerstr"; })");
+  test("BEGIN { let $a: int16 = 1; $a = (int8)2; }");
+  // Test more types
+  test("BEGIN { let $a: struct x; }");
+  test("BEGIN { let $a: struct x *; }");
+
+  test_error("BEGIN { let $a; let $a; }", R"(
+stdin:1:17-23: ERROR: Variable $a was already declared.
+BEGIN { let $a; let $a; }
+                ~~~~~~
+stdin:1:9-15: ERROR: Initial declaration
+BEGIN { let $a; let $a; }
+        ~~~~~~
+)");
+
+  test_error("BEGIN { let $a: uint16; $a = -1; }", R"(
+stdin:1:26-33: ERROR: Type mismatch for $a: trying to assign value of type 'int64' when variable already has a type 'uint16'
+BEGIN { let $a: uint16; $a = -1; }
+                         ~~~~~~~
+)");
+
+  test_error("BEGIN { let $a; $a = (uint8)1; $a = -1; }", R"(
+stdin:1:32-39: ERROR: Type mismatch for $a: trying to assign value of type 'int64' when variable already contains a value of type 'uint8'
+BEGIN { let $a; $a = (uint8)1; $a = -1; }
+                               ~~~~~~~
+)");
+
+  test_error("BEGIN { let $a: int8; $a = 10000; }", R"(
+stdin:1:24-34: ERROR: Type mismatch for $a: trying to assign value '10000' which does not fit into the variable of type 'int8'
+BEGIN { let $a: int8; $a = 10000; }
+                       ~~~~~~~~~~
+)");
+
+  test_error("BEGIN { $a = -1; let $a; }", R"(
+stdin:1:18-24: ERROR: Variable declarations need to occur before variable usage or assignment. Variable: $a
+BEGIN { $a = -1; let $a; }
+                 ~~~~~~
+)");
+
+  test_error("BEGIN { let $a: uint16 = -1; }", R"(
+stdin:1:9-29: ERROR: Type mismatch for $a: trying to assign value of type 'int64' when variable already has a type 'uint16'
+BEGIN { let $a: uint16 = -1; }
+        ~~~~~~~~~~~~~~~~~~~~
+)");
+
+  test_error(R"(BEGIN { let $a: str_t[5] = "hiya"; $a = "longerstr"; })", R"(
+stdin:1:37-53: ERROR: Type mismatch for $a: trying to assign value of type 'string[10]' when variable already contains a value of type 'string[5]'
+BEGIN { let $a: str_t[5] = "hiya"; $a = "longerstr"; }
+                                    ~~~~~~~~~~~~~~~~
+)");
+
+  test_error(R"(BEGIN { let $a; print(($a)); $a = 1; })", R"(
+stdin:1:17-26: ERROR: Variable used before it was assigned: $a
+BEGIN { let $a; print(($a)); $a = 1; }
+                ~~~~~~~~~
+)");
+
+  test_error(R"(BEGIN { let $a: sum_t; })", R"(
+stdin:1:9-23: ERROR: Invalid variable declaration type: sum
```

Comment:
Shouldn't be
```suggestion
stdin:1:9-23: ERROR: Invalid variable declaration type: sum_t
```

Reply:
arrrgghhhh - the only time these types are available is in the parser. Our `typestr` function in types.cpp doesn't have any `_t` postfix for any types.